### PR TITLE
Find merchant total revenue

### PR DESCRIPTION
**Pull request by TheKaterTot**
Finds the total revenue for a specific merchant.

### CODE REVIEW

File: app/models/invoice.rb [hunk 1]
```diff
@@ -3,6 +3,19 @@ class Invoice < ApplicationRecord
...
 
   validates :status, presence: true
+
+  def total
+    invoice_items.reduce(0) do |total, item|
+      total += (item.unit_price * item.quantity)
+    end
+  end
+
+  def successful?
```
Comment (TheKaterTot) [inline app/models/invoice.rb:16]: These methods are here to support the query. Should I be doing this as part of the query @s-espinosa?

File: app/models/merchant.rb [hunk 2]
```diff
@@ -4,4 +4,14 @@ class Merchant < ApplicationRecord
...
 
   validates :name, presence: true
+
+  def total_revenue
+    invoices.select(&:successful?).reduce(0) do |total, invoice|
+      total += invoice.total
+    end
+  end
+
+  def self.with_invoices
```
Comment (TheKaterTot) [inline app/models/merchant.rb:14]: @s-espinosa Here are the methods that are doing the calculations as well as doing some querying. Is this sufficient or this too much Ruby for your taste?

File: app/models/invoice.rb [hunk 3]
```diff
@@ -3,6 +3,19 @@ class Invoice < ApplicationRecord
   belongs_to :merchant
 
   has_many :transactions
+  has_many :invoice_items
```
Comment (CjMoore) [inline app/models/invoice.rb:6]: good catch. 👍

**Review by CjMoore — APPROVED** [review]
Looks good 👍

File: app/models/merchant.rb (hunk 2 above)
Comment (s-espinosa) [reply app/models/merchant.rb:14]: I like the ideas here, but do recommend more AR and less vanilla Ruby. It seems to me that what you're trying to return is a collection of Invoice objects that have a successful transaction. You can start with that `includes` call, and then scope the result down using `.where`.

Remember, `.where` can take raw SQL as an argument, so you should be able to specify a particular result on `transactions` (there's also a way to do this with a nested hash, but the SQL might be more straightforward?). We can definitely talk through this more tomorrow.

File: app/models/invoice.rb (hunk 1 above)
Comment (s-espinosa) [reply app/models/invoice.rb:16]: Depends on what you mean. I like the idea of having some of this in separate methods that can then be reused, but I think it's possible to do that (maybe using a scope?) while still using only ActiveRecord.

**Review by s-espinosa — COMMENTED** [review]
@TheKaterTot Good questions. My comments below. 👍

File: app/models/invoice.rb (hunk 1 above)
Comment (TheKaterTot) [reply app/models/invoice.rb:16]: Moving successful results feels tricky to me. If I do it as part my query in merchants, I'll be filtering merchants. I want to make sure that I'm leaving that logic with the object it belongs with, right?

Comment (s-espinosa) [reply app/models/invoice.rb:16]: @TheKaterTot I think that's exactly right! Let's talk more tomorrow, but I believe that `success` is fundamentally a characteristic of transactions. By proxy, it's (to *some* degree) a characteristic of invoices and invoice_items, but moving it too far away from those models feels strange to me (personal opinion, for sure, and could even see an argument that those last two are too far).

We'll discuss this tomorrow, but you might want to take a look at `.merge`. Essentially, it allows you to scope your join down to some subset of the joined model using a class method on that model.

**Pull request merged**